### ISSUE TITLE
Installation on Mac OS (version 14.2)

### ISSUE THREAD

**Issue by vabor112** [wesselb/lab#15]
Hey @wesselb! Someone from [GeometricKernels](https://github.com/GPflow/GeometricKernels) over here.

I was trying to install the library on a new mac following your [instructions](https://gist.github.com/wesselb/4b44bf87f3789425f96e26c4308d0adc), using `pip`. However, I got these errors:
`gcc-13: error: unrecognized command-line option '-iwithsysroot/System/Library/Frameworks/System.framework/PrivateHeaders'`   
`gcc-13: error: unrecognized command-line option '-iwithsysroot/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/Headers'`

These `-iwithsysroot` are nothing more than Clang syntax for `-isysroot`. However, digging into it a bit, I figured that they (1) are added to `CFLAGS` by Python's distutils (like [here](https://stackoverflow.com/questions/68433956/building-python-3-c-modules-in-macos-setup-using-odd-compiler-options)) and (2) are super hard to get rid of.

Unsuccessful at using gcc, I found that there's actually a way to use mac's default Clang with OpenMP. It is described [here](https://mac.r-project.org/openmp/). So I removed the part of `setup.py` that changes `CC` to gcc and changed `"-fopenmp"` option to `"-Xclang", "-fopenmpp"`. This stopped the `unrecognized command-line option` errors.

The comiler was still having trouble finding fortran after that so I had to additionally add `"-L/opt/homebrew/Cellar/gcc/13.2.0/lib/gcc/current"` to the `extra_link_args`. After this, I was able to sucessfully install the library.

What are your thoughts? Maybe you could update `setup.py` and instructions accordingly? The only thing that requires a bit more work, it seems, is making `-L/opt/homebrew/Cellar/gcc/13.2.0/lib/gcc/current` into something more generic. Maybe you have ideas right away. If not, I can also think about how to go about this.

Here's the setup.py I ended up with:
```python
import os
import subprocess

import numpy as np
from Cython.Build import build_ext
from setuptools import Extension, find_packages, setup

if os.name == "nt":
    # On Windows, don't compile the extra modules.
    ext_modules = []

elif "LAB_NO_BUILD" in os.environ and os.environ["LAB_NO_BUILD"] == "1":
    # Also don't build anything if `LAB_NO_BUILD` is set to `1`.
    ext_modules = []

else:
    # Include libraries from the OS X Command Line Tools. On OS X Big Sur, these
    # libraries are not automatically included anymore.
    osx_library_path = "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib"
    if os.path.exists(osx_library_path):
        if "LIBRARY_PATH" in os.environ and os.environ["LIBRARY_PATH"]:
            os.environ["LIBRARY_PATH"] += ":" + osx_library_path
        else:
            os.environ["LIBRARY_PATH"] = osx_library_path

    # If `xcrun` is available, make sure the includes are added to CPATH.
    if subprocess.call("which xcrun", shell=True) == 0:
        path = (
            subprocess.check_output("xcrun --show-sdk-path", shell=True)
            .strip()
            .decode("ascii")
        )
        path += "/usr/include"

        # Add to CPATH.
        if "CPATH" not in os.environ:
            os.environ["CPATH"] = ""
        os.environ["CPATH"] += path

    # Default to use gcc as the compiler if `$CC` is not set.
    if "CC" not in os.environ or not os.environ["CC"]:
        os.environ["CC"] = "gcc"

    # Check whether `gfortran` is available.
    if subprocess.call("which gfortran", shell=True) != 0:
        if "LAB_GFORTRAN" in os.environ and os.environ["LAB_GFORTRAN"]:
            gfortran = os.environ["LAB_GFORTRAN"]
        else:
            gfortran = False
    else:
        gfortran = "gfortran"

    # Compile TVPACK if `gfortran` is available.
    if gfortran:
        if (
            subprocess.call(
                f"{gfortran} -fPIC -O2 -c lab/bvn_cdf/tvpack.f -o lab/bvn_cdf/tvpack.o",
                shell=True,
            )
            != 0
        ):
            raise RuntimeError("Compilation of TVPACK failed.")

    # Determine which external modules to compile.
    ext_modules = []

    if gfortran:
        extra_objects = ["lab/bvn_cdf/tvpack.o"]
        extra_link_args = ["-Xclang", "-fopenmpp", "-L/opt/homebrew/Cellar/gcc/13.2.0/lib/gcc/current"]

        # Allow the libraries for `gfortran` to be explicitly linked.
        if "LAB_LIBGFORTRAN" in os.environ and os.environ["LAB_LIBGFORTRAN"]:
            extra_objects += [os.environ["LAB_LIBGFORTRAN"]]
        else:
            extra_link_args += ["-lgfortran"]

        ext_modules.append(
            Extension(
                "lab.bvn_cdf",
                sources=["lab/bvn_cdf/bvn_cdf.pyx"],
                include_dirs=[np.get_include()],
                extra_compile_args=["-fPIC", "-O2"],
                extra_objects=extra_objects,
                extra_link_args=extra_link_args,
            )
        )

requirements = [
    "numpy>=1.16",
    "scipy>=1.3",
    "fdm",
    "plum-dispatch>=2",
    "opt-einsum",
]

setup(
    packages=find_packages(exclude=["docs"]),
    python_requires=">=3.8",
    install_requires=requirements,
    cmdclass={"build_ext": build_ext},
    ext_modules=ext_modules,
    include_package_data=True,
)
```

**Comment by wesselb** [comment]
Hey @vabor112! Thanks for opening an issue. I've played a little with trying to use `clang` for the compilation, but on my end I can't get it to recognise the flag` -fopenmp`. Everything online seems to suggest to use `brew`s `gcc` or `llvm`. :(

I think your suggestion of not using macOS's system `python` might be the best way around this. I've added a note to the installation instructions.

By the way, compilation can be avoided all together by setting `LAB_NO_BUILD=1`. Since compilation is only needed for very specific features, I think I should compilation by default, and let the user opt to to compilation whenever they want to make use of the very specific features...

**Comment by vabor112** [comment]
Thank you @wesselb for the quick response!

> I've played a little with trying to use clang for the compilation, but on my end I can't get it to recognise the flag -fopenmp. 

The trick with `-fopenmp`, as R project developers [have discovered](https://mac.r-project.org/openmp/), is that Apple removed this flag from their Clang in a very sloppy way. Although `-fopenmp` does not work, the combination `-Xclang -fopenmp` does. And here `-Xclang ...` is an option which basically says "pass ... as an option to Clang". I did not need anything else besides just switching from `-fopenmp` to `-Xclang -fopenmp`.

> I think your suggestion of not using macOS's system python might be the best way around this. I've added a note to the installation instructions.

Thanks!

> By the way, compilation can be avoided all together by setting LAB_NO_BUILD=1. Since compilation is only needed for very specific features, I think I should compilation by default, and let the user opt to to compilation whenever they want to make use of the very specific features...

Wow, I did not know about that! It makes total sense to turn off compilation by default. I am not sure though if we use any of these specific features requiring compilation in GeometricKernels. Where can I find out what these are, exactly?

Feel free to close this issue!

**Comment by wesselb** [comment]
@vabor112, ah you're completely right! First passing `-Xclang` works. I think `-L/opt/homebrew/Cellar/gcc/13.2.0/lib/gcc/current` could be automatically extracted from `gcc-13 -print-search-dirs`. If this is more robust, then I'm happy to change it that way.


> Wow, I did not know about that! It makes total sense to turn off compilation by default. I am not sure though if we use any of these specific features requiring compilation in GeometricKernels. Where can I find out what these are, exactly?

It's actually just one feature: `B.bvn_cdf`. 😅 I should've turned off compilation by default from the very beginning. I will change that immediately and release a new version.

**Comment by wesselb** [comment]
Hey @vabor112! I've released a new version `v1.6.0` which turns off any compilation. LAB should now be a lightweight dependency! I'm sorry for the trouble this has caused in the first place.

Will close the issue for now as I think this should solve, but please comment and/or reopen if you continue to experience difficulties.
